### PR TITLE
Improve the `typeinfo` module

### PR DESCRIPTION
**Pull request by konsumlamm**
* add example
* improve wording
* use `let` instead of `var` in some places

There are some warnings (for example `ze`), but I'm not really sure how to fix them.

## future work
- [x] remove uses of ze in this module refs https://github.com/nim-lang/Nim/pull/17625#pullrequestreview-627549215
- [ ] add a generic API for getters and setters refs https://github.com/nim-lang/Nim/pull/17625#pullrequestreview-627569520

### CODE REVIEW

File: lib/core/typeinfo.nim [hunk 1]
```diff
@@ -120,7 +133,7 @@ template `+!!`(a, b): untyped = cast[pointer](cast[ByteAddress](a) + b)
 proc getDiscriminant(aa: pointer, n: ptr TNimNode): int =
   assert(n.kind == nkCase)
   var d: int
-  var a = cast[ByteAddress](aa)
+  let a = cast[ByteAddress](aa)
   case n.typ.size
   of 1: d = ze(cast[ptr int8](a +% n.offset)[])
   of 2: d = ze(cast[ptr int16](a +% n.offset)[])
```
Comment (metagn) [inline lib/core/typeinfo.nim:139]:
```suggestion
  of 1: d = int(cast[ptr uint8](a +% n.offset)[])
  of 2: d = int(cast[ptr uint16](a +% n.offset)[])
```

I believe this is how you change `ze`. Also `d` in this proc could just be `result` I think.

File: lib/core/typeinfo.nim [hunk 2]
```diff
@@ -144,17 +157,17 @@ proc newAny(value: pointer, rawType: PNimType): Any {.inline.} =
 
 when declared(system.VarSlot):
   proc toAny*(x: VarSlot): Any {.inline.} =
-    ## Constructs a `Any` object from a variable slot `x`.
+    ## Constructs an `Any` object from a variable slot `x`.
     ## This captures `x`'s address, so `x` can be modified with its
-    ## `Any` wrapper! The client needs to ensure that the wrapper
+    ## `Any` wrapper! The user needs to ensure that the wrapper
```
Comment (timotheecour) [inline lib/core/typeinfo.nim]:
```suggestion
    ## `Any` wrapper! The caller needs to ensure that the wrapper
```

ditto elsewhere

File: lib/core/typeinfo.nim [hunk 3]
```diff
@@ -167,12 +180,12 @@ proc size*(x: Any): int {.inline.} =
   result = x.rawType.size
 
 proc baseTypeKind*(x: Any): AnyKind {.inline.} =
-  ## Gets the base type's kind; `akNone` is returned if `x` has no base type.
+  ## Gets the base type's kind. If `x` has no base type,`akNone` is returned.
```
Comment (timotheecour) [inline lib/core/typeinfo.nim]:
```suggestion
  ## Gets the base type's kind. If `x` has no base type, `akNone` is returned.
```

File: lib/core/typeinfo.nim [hunk 4]
```diff
@@ -301,33 +314,32 @@ proc len*(x: Any): int =
...
-                            tySequence, tyProc}
+const pointerLike =
+  when defined(gcDestructors): {tyCString, tyRef, tyPtr, tyPointer, tyProc}
+  else: {tyString, tyCString, tyRef, tyPtr, tyPointer, tySequence, tyProc}
 
 proc getPointer*(x: Any): pointer =
   ## Retrieves the pointer value out of `x`. `x` needs to be of kind
   ## `akString`, `akCString`, `akProc`, `akRef`, `akPtr`,
-  ## `akPointer`, `akSequence`.
+  ## `akPointer` or `akSequence`.
```
Comment (timotheecour) [inline lib/core/typeinfo.nim:335]: pre-existing but:
* the comment and the assert seem to say 2 different things (assert is for x.kind, comment is for x.rawType.kind)
* `when defined(gcDestructors)` affects pointerLike but doc comment doesn't reflect this
* repeating constraint in doc comment and assert is not DRY

## proposal
via https://github.com/nim-lang/RFCs/issues/270
```nim
proc getPointer*(x: Any): pointer {.requires: x.rawType.kind in pointerLike.} =
  ## Retrieves the pointer value out of `x`.
  result = cast[ppointer](x.value)[]
```
=> self documenting, DRY, and helps static analyzers

(obviously, out of scope for this PR, but just wanted to mention this)

File: lib/core/typeinfo.nim [hunk 5]
```diff
@@ -505,38 +516,34 @@ proc setBiggestInt*(x: Any, y: BiggestInt) =
...
   result = cast[ptr uint16](x.value)[]
 
 proc getUInt32*(x: Any): uint32 =
-  ## Retrieves the uint32 value out of `x`, `x` needs to represent an
-  ## uint32.
+  ## Retrieves the `uint32` value out of `x`. `x` needs to represent a `uint32`.
   assert skipRange(x.rawType).kind == tyUInt32
   result = cast[ptr uint32](x.value)[]
 
 proc getUInt64*(x: Any): uint64 =
```
Comment (timotheecour) [inline lib/core/typeinfo.nim:538]: pre-existing but these API's are bad; we should offer generic getter's/setters, eg:

```nim
proc get*(x: Any, T: typedesc): T =
  ...
```
followup PR welcome, if possible

File: lib/core/typeinfo.nim [hunk 6]
```diff
@@ -672,10 +679,10 @@ proc assign*(x, y: Any) =
   genericAssign(x.value, y.value, y.rawType)
 
 iterator elements*(x: Any): int =
-  ## Iterates over every element of `x` that represents a Nim bitset.
+  ## Iterates over every element of `x`. `x` needs to represent a Nim bitset.
```
Comment (timotheecour) [inline lib/core/typeinfo.nim]:
```suggestion
  ## Iterates over every element of `x`. `x` needs to represent a `set`.
```

File: lib/core/typeinfo.nim (hunk 1 above)
Comment (timotheecour) [reply lib/core/typeinfo.nim:139]: yes but let's defer this to followup PR

File: lib/core/typeinfo.nim [hunk 7]
```diff
@@ -144,17 +157,17 @@ proc newAny(value: pointer, rawType: PNimType): Any {.inline.} =
...
     ## **does not** live longer than `x`!
     ## This is provided for easier reflection capabilities of a debugger.
     result.value = x.address
     result.rawType = x.typ
 
 proc toAny*[T](x: var T): Any {.inline.} =
-  ## Constructs a `Any` object from `x`. This captures `x`'s address, so
-  ## `x` can be modified with its `Any` wrapper! The client needs to ensure
+  ## Constructs an `Any` object from `x`. This captures `x`'s address, so
+  ## `x` can be modified with its `Any` wrapper! The user needs to ensure
```
Comment (timotheecour) [inline lib/core/typeinfo.nim]:
```suggestion
  ## `x` can be modified with its `Any` wrapper! The caller needs to ensure
```

**Review by timotheecour — APPROVED** [review]

**Pull request merged**